metrics.go

```diff
@@ -20,16 +20,91 @@ directly to Stackdriver Metrics.
 */
 
 import (
+	"errors"
 	"fmt"
+	"path"
 
 	"github.com/golang/protobuf/ptypes/timestamp"
 
 	distributionpb "google.golang.org/genproto/googleapis/api/distribution"
+	labelpb "google.golang.org/genproto/googleapis/api/label"
+	googlemetricpb "google.golang.org/genproto/googleapis/api/metric"
 	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
 
 	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
 )
 
+var errNilMetric = errors.New("expecting a non-nil metric")
+
+func (se *statsExporter) protoToMonitoringMetricDescriptor(metric *metricspb.Metric) (*googlemetricpb.MetricDescriptor, error) {
+	if metric == nil {
+		return nil, errNilMetric
+	}
+
+	metricName, description, unit, _ := metricProseFromProto(metric)
+	metricType, _ := se.metricTypeFromProto(metricName)
+	displayName := se.displayName(metricName)
+	metricKind, valueType := protoMetricDescriptorTypeToMetricKind(metric)
+
+	sdm := &googlemetricpb.MetricDescriptor{
+		Name:        fmt.Sprintf("projects/%s/metricDescriptors/%s", se.o.ProjectID, metricType),
+		DisplayName: displayName,
+		Description: description,
+		Unit:        unit,
+		Type:        metricType,
+		MetricKind:  metricKind,
+		ValueType:   valueType,
+		Labels:      labelDescriptorsFromProto(se.defaultLabels, metric.GetMetricDescriptor().GetLabelKeys()),
+	}
+
+	return sdm, nil
+}
+
+func labelDescriptorsFromProto(defaults map[string]labelValue, protoLabelKeys []*metricspb.LabelKey) []*labelpb.LabelDescriptor {
+	labelDescriptors := make([]*labelpb.LabelDescriptor, 0, len(defaults)+len(protoLabelKeys))
+
+	// Fill in the defaults first.
+	for key, lbl := range defaults {
+		labelDescriptors = append(labelDescriptors, &labelpb.LabelDescriptor{
+			Key:         sanitize(key),
+			Description: lbl.desc,
+			ValueType:   labelpb.LabelDescriptor_STRING,
+		})
+	}
+
+	// Now fill in those from the metric.
+	for _, protoKey := range protoLabelKeys {
+		labelDescriptors = append(labelDescriptors, &labelpb.LabelDescriptor{
+			Key:         sanitize(protoKey.GetKey()),
+			Description: protoKey.GetDescription(),
+			ValueType:   labelpb.LabelDescriptor_STRING, // We only use string tags
+		})
+	}
+	return labelDescriptors
+}
+
+func metricProseFromProto(metric *metricspb.Metric) (name, description, unit string, ok bool) {
```

Comment:
Is "Prose" a typo?

Reply:
No, "Prose" means ordinary words basically simple words and in this case we are describing the metric so name, description, unit.